[[ Bug 21051 ]] Handle wchar_t size differences for ODBC

montegoulding opened the pull request:
This patch works around `wchar_t` size difference on unix machines which
was causing unicode results from ODBC databases to be truncated after the
first character due to null bytes.

runrevmark commented:
@montegoulding Can we determine whether the output of ODBC is UTF-16 just being put into 32-bit ints, or whether it is actually UTF-32?

montegoulding commented:
Good question @runrevmark ... my googlefu failed me this afternoon trying to find that out

montegoulding commented:
Aha! found it. Assuming Patrick knows what he's talking about then my patch is wrong and I need to convert from UTF32 properly https://sourceforge.net/p/iodbc/mailman/message/1028013/

runrevmark commented:
Heh - I shall refrain from comment about the ridiculousness of UTF-32/UCS-4 but anyway... It is easy enough to test - shove a unicode char with code > 65535 in to a field and see if you get that value back, or two values < 65535. Fortunately the conversion from 32->16 is easy enough and you can do it in place (a single UTF-32 'char' takes at most 2 UTF-16 'chars').

montegoulding committed:
[[ Bug 21051 ]] Handle wchar_t size differences for ODBC

This patch works around `wchar_t` size difference on unix machines which
was causing unicode results from ODBC databases to be truncated after the
first character due to null bytes.

montegoulding committed:
[[ Bug ]] Fix ODBC parameter binding

This patch fixes ODBC parameter binding by correcting the use of the
`StrLen_or_IndPtr` parameter which should point to a buffer containing
the length of the parameter value and that is valid when `SQLExecute` is
called.

livecodemichael commented:
@livecode-vulcan review ok a0eec38

livecode-vulcan commented:
💙 review by @livecodemichael ok a0eec38dde57cd87545123d46b5bdbed1a9d2bec

livecode-vulcan commented:
😞 test failure a0eec38dde57cd87545123d46b5bdbed1a9d2bec

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/1462)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/1478)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/1452)
* try-community-universal-ios-iphoneos11.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos11.3/builds/55)
* try-community-universal-ios-iphonesimulator11.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator11.3/builds/55)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/1456)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/1482)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/1462)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/1476)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/1461)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/1471)
* try-community-x86-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/1511)
* try-community-x86_64-win32: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/1451)

livecodepanos commented:
@livecode-vulcan retry a0eec38

livecode-vulcan commented:
😎 test success a0eec38dde57cd87545123d46b5bdbed1a9d2bec

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/1465)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/1481)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/1455)
* try-community-universal-ios-iphoneos11.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos11.3/builds/58)
* try-community-universal-ios-iphonesimulator11.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator11.3/builds/58)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/1459)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/1485)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/1465)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/1479)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/1464)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/1474)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/1514)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/1454)
